features/popup_chat.lua

```diff
@@ -0,0 +1,76 @@
+local Event = require 'utils.event'
+local Global = require 'utils.global'
+
+local config = global.config.popup_chat
+local MIN_LIFETIME = config.min_lifetime or 06 * 60 -- 06s
+local MAX_LIFETIME = config.max_lifetime or 20 * 60 -- 20s
+local MIN_MESSAGE_LENGTH = 40
+local MAX_MESSAGE_LENGTH = 92
+local TIME_PER_CHAR = 3 -- about +1 sec every 20 chars (60/20 ticks/chars)
+
+local data = {
+  popup_chat = {}
+}
+
+Global.register(data, function(tbl)
+  data = tbl
+end)
+
+---@param message string
+local function message_lifetime(message)
+  local length = message:len()
+  if length <= MIN_MESSAGE_LENGTH then
+    return MIN_LIFETIME
+  end
+  local extra_time = math.floor((length - MIN_MESSAGE_LENGTH) * TIME_PER_CHAR)
+  return math.min(MIN_LIFETIME + extra_time, MAX_LIFETIME)
+end
+
+---@param message string
+local function safe_message(message)
+  local length = message:len()
+  if length <= MAX_MESSAGE_LENGTH then
+    return message
+  end
+  return string.sub(message, 1, MAX_MESSAGE_LENGTH) .. '[...]'
+end
+
+---@param event defines.event.on_console_chat
+local function on_console_chat(event)
+  local index = event.player_index
+  local message = event.message
+  if not (index and message) then
+    return
+  end
+
+  local player = game.players[event.player_index]
+  if not (player and player.character) then
```

Comment:
```suggestion
  if not (player and player.valid and player.character) then
```
Very unlikely but technically player can be not valid.